modules/RTC/CodecSelection.js

```diff
@@ -21,40 +21,67 @@ export class CodecSelection {
-     * Falling back to the standard codec will be skipped when this option is true, endpoints
-     * that do not support the preferred codec may not be able to encode/decode video when this happens.
-     * @param {string} options.jvbCodec the codec that is preferred on jvb connection.
-     * @param {string} options.p2pCodec the codec that is preferred on p2p connection.
+     * @param {string} options.jvb settings (codec list, preferred and disabled) for the jvb connection.
+     * @param {string} options.p2p settings (codec list, preferred and disabled) for the p2p connection.
      */
     constructor(conference, options) {
         this.conference = conference;
         this.options = options;
-        this.enforcePreferredCodec = options.enforcePreferredCodec;
-
-        // VP8 cannot be disabled since it the default codec.
-        this.p2pDisabledCodec = options.p2pDisabledCodec !== CodecMimeType.VP8
-            && this._isCodecSupported(options.p2pDisabledCodec)
-            && options.p2pDisabledCodec;
-        this.jvbDisabledCodec = options.jvbDisabledCodec !== CodecMimeType.VP8
-            && this._isCodecSupported(options.jvbDisabledCodec)
-            && options.jvbDisabledCodec;
-
-        // Determine the preferred codecs.
-        this.p2pPreferredCodec = this._isCodecSupported(options.p2pPreferredCodec)
-            && options.p2pPreferredCodec !== options.p2pDisabledCodec
-            ? options.p2pPreferredCodec
-            : CodecMimeType.VP8;
-        this.jvbPreferredCodec = this._isCodecSupported(options.jvbPreferredCodec)
-            && options.jvbPreferredCodec !== options.jvbDisabledCodec
-            ? options.jvbPreferredCodec
-            : CodecMimeType.VP8;
-
-        logger.debug(`Codec preferences for the conference are JVB: preferred=${this.jvbPreferredCodec},`
-            + `disabled=${this.jvbDisabledCodec} P2P: preferred=${this.p2pPreferredCodec},`
-            + `disabled=${this.p2pDisabledCodec}`);
+        this.codecPreferenceOrder = {};
+
+        for (const connectionType of Object.keys(options)) {
+            // eslint-disable-next-line prefer-const
+            let { disabledCodec, preferredCodec, preferenceOrder } = options[connectionType];
+            const supportedCodecs = new Set(this._getSupportedVideoCodecs());
+            let selectedOrder;
+
+            if (preferenceOrder) {
+                // Select all codecs that are supported by the browser.
+                selectedOrder = preferenceOrder
+                    .filter(codec => supportedCodecs.has(codec.toLowerCase()));
+
+                if (!browser.supportsVP9()) {
+                    const index = selectedOrder.findIndex(codec => codec.toLowerCase() === CodecMimeType.VP9);
+
+                    if (index !== -1) {
+                        selectedOrder.splice(index, 1);
+                        selectedOrder.push(CodecMimeType.VP9);
```

Comment:
Maybe a comment here, indicating that we are putting VP9 last so at least decoding works would help future readers.